content/blog/2022-05-18-learned-federated-query-optimization.md

```diff
@@ -0,0 +1,178 @@
+evaluate the query and get the resulting rows.
+
+### Compilation Lifecycle
+The steps of compilation of a query consists of the following steps.
+1. Parsing: We use Spark SQL query parser
+2. Intermediate Representation: The first intermediate representation is a tree-like structure, where each parent node represents the join between two tables
+3. Graph Representation: We tranform the aforementioned intermediate representation to a more modular graph structure, which gives us more freedom on create efficient plans.
+4. Optimization: By examining the graph, we produce a the optimal (according to the cost model) tree-like query execution plan
+5. Execution: We pass the final plan back to the federation engine for evaluation
+
+We explain in more detail the intermediate representation that our compiler uses.
+
+### Graph Intermediate Representation
+Let's take again a look at the previous example
+```sql
+SELECT AVG(A.salary)
+FROM A, B, C
+WHERE A.id = B.id
+AND A.id = C.id
+AND B.id = C.id
+```
+
+If we think of this query as a graph, it would look as follows:
+```
+    A - B
+    \   / 
+      C
+```
+
+This is a fully-connected graph, as there are connections from all nodes to all nodes. As the final plan will be a binary tree,
+that means that the final plan could be any of the following three trees:
+```
+1.  ⋈         2.  ⋈          3.  ⋈ 
+   /  \          /  \           /  \
+  A   ⋈         B   ⋈          C   ⋈ 
+     / \           /  \           /  \
+    B   C         A    C         A    B
+```
+
+However, we are mostly interested in plans that look like 1. The reason is that we are mainly interested in joins that 
+can co-locate tables thus, so execution push-down is possible. Thus, we created a graph structure in which each node
+represents a table, and each edge represents a join between two tables. We enrich more our graph representation by adding
+information about the location of each table in the node, that is, in which data source the node resides. This allows
+us to develop an algorithm can can parse this graph and generate possible trees that join tables that reside in the same
+engine, when that's possible given the input query semantics (joins).
+
+### Final Plan Generation
+After processing the graph, we get a tree that collocates tables residing in the same engine by producing a tree-plan
+that joins together the collocated tables. Then, we use our machine learning model in order to predict the execution cost
+of joining the collocated tables in the machine that they reside on in the federation engine. We then produce the final
+query execution plan which also contains information about where each node of the tree will be executed.
+
+# What were the hardest parts to get right?
+The trickiest part of the project, which is still not completely resolved is the cost estimation. The average accuracy of 
+our model in predicting the cost of a given query is around 60%. However, in most of the cases in which we mispredict
```

Comment:
Define what you mean by "accuracy." L1 norm, normalized to some maximum cost, maybe?